### PR TITLE
Chet 494 retry provisioning

### PR DESCRIPTION
**Pull request by theghostwhoforks**

### CODE REVIEW

File: api/src/main/kotlin/com/atlassian/migration/datacenter/api/aws/CloudFormationEndpoint.kt [hunk 1]
```diff
@@ -102,5 +102,21 @@ class CloudFormationEndpoint(private val deploymentService: ApplicationDeploymen
...
+    @POST
+    @Path("/reset")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    fun resetProvisioningStage(): Response {
+        return when (val currentStage = migrationService.currentStage) {
+            MigrationStage.ERROR, //Only until we change transitioning to errors in provisioning steps
+            MigrationStage.PROVISIONING_ERROR -> {
+                migrationService.transition(MigrationStage.PROVISION_APPLICATION)
+                Response.accepted().build()
```
Comment (bpar476) [inline api/src/main/kotlin/com/atlassian/migration/datacenter/api/aws/CloudFormationEndpoint.kt:114]: Should this be `ok` instead of `accepted`? There isn't any asynchronous work running after this request returns.

File: api/src/test/kotlin/com/atlassian/migration/datacenter/api/MigrationEndpointTest.kt [hunk 2]
```diff
@@ -57,6 +57,24 @@ class MigrationEndpointTest {
         assertThat(response.entity.toString(), Matchers.containsString(MigrationStage.AUTHENTICATION.toString()))
     }
 
+    @Test
+    fun shouldBeErrorWhenStageIsASpecificErrorStage() {
+        every { migrationService.currentStage } returns MigrationStage.FINAL_SYNC_ERROR
+
+        val response = sut.getMigrationStatus()
+
+        assertThat(response.entity.toString(), Matchers.containsString(MigrationStage.ERROR.toString()))
```
Comment (bpar476) [inline api/src/test/kotlin/com/atlassian/migration/datacenter/api/MigrationEndpointTest.kt:66]: Why do we want to obfuscate the specific error stage in the API?

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt [hunk 3]
```diff
@@ -19,56 +19,83 @@ package com.atlassian.migration.datacenter.spi
  * Represents all possible states of an on-premise to cloud migration.
  */
 enum class MigrationStage {
-    NOT_STARTED,
+    NOT_STARTED(),
     AUTHENTICATION(NOT_STARTED),
+
     PROVISION_APPLICATION(AUTHENTICATION),
```
Comment (bpar476) [inline spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:25]: Does this need to be valid from `PROVISIONING_ERROR` as well?

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt [hunk 4]
```diff
@@ -19,56 +19,83 @@ package com.atlassian.migration.datacenter.spi
...
+        }
     }
 
     fun isAfter(stage: MigrationStage): Boolean {
         if (this == stage || this == NOT_STARTED) return false
         if (this == FINISHED || this == ERROR) return true
-        this.validFrom!!.let { it: MigrationStage ->
-            return if (it == stage) true else it.isAfter(stage)
-        }
+        return this.validFromStages.map {
```
Comment (bpar476) [inline spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:82]: can this not just be `validFromStages.contains(stage)`

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt [hunk 5]
```diff
@@ -19,56 +19,83 @@ package com.atlassian.migration.datacenter.spi
...
+        }.contains(true)
     }
 
     // Hacky, but OK for now.
     val isDBPhase: Boolean
         get() =// Hacky, but OK for now.
             this.toString().startsWith("DB_")
 
+    fun isErrorStage(): Boolean {
+        return listOf(ERROR, FS_MIGRATION_ERROR, PROVISIONING_ERROR, FINAL_SYNC_ERROR).contains(this)
```
Comment (bpar476) [inline spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:96]: This is a very minor thing but would this be better as
```
return when(this) {
     ERROR, FS_MIGRATION_ERROR, PROVISIONING_ERROR, FINAL_SYNC_ERROR -> true
     else -> false
}
```
This way we don't need to construct a list object or iterate over it

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt [hunk 6]
```diff
@@ -19,56 +19,83 @@ package com.atlassian.migration.datacenter.spi
...
+    constructor(vararg validFromStages: MigrationStage) {
+        exception = null;
+        this.validFromStages = validFromStages.asList()
     }
 
+
     fun isValidTransition(to: MigrationStage): Boolean {
-        return to.validFrom?.equals(this) ?: true
+        when (this) {
+            PROVISIONING_ERROR -> return to == PROVISION_APPLICATION
```
Comment (bpar476) [inline spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: If we add `PROVISIONING_ERROR` to the `validFromStages` of `PROVISION_APPLICATION` we don't need to have this edge case.

File: api/src/test/kotlin/com/atlassian/migration/datacenter/api/MigrationEndpointTest.kt (hunk 2 above)
Comment (theghostwhoforks) [reply api/src/test/kotlin/com/atlassian/migration/datacenter/api/MigrationEndpointTest.kt:66]: This was done just to maintain compatibility with the currrent UI.

When we play the UI story, I'll relook at the stage specific errors and see if they need to be handled

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt (hunk 6 above)
Comment (theghostwhoforks) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: That causes a circular dependency when declaring both `PROVISIONING_ERROR(PROVISION_APPLICATION)` and `PROVISIONING_APPLICATION(PROVISIONING_ERROR)`

We need to pick either or :(

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt (hunk 4 above)
Comment (theghostwhoforks) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:82]: In this case, yes. But the recursion was easier to model in terms of nesting in each element of the `validFromStages` list

  A
 |-|
B  C
----|
      D

Here, with recursion D.isAfter(A) will evaluate to true as we check C's ancestors.

With `validFromStages.contains(stage)`, we would stop at `C` and not traverse to `A`.

Does this make sense of have I misunderstood something?

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt (hunk 3 above)
Comment (theghostwhoforks) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:25]: see below

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt (hunk 6 above)
Comment (bpar476) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: Retrying is inherently a cycle and if you were to draw the state machine as a graph there would indeed be a cycle between PROVISION_APPLICATION to PROVISIONING_ERROR.

Where exactly does the circular dependency cause an error. Is it just the `isAfter` functionality?

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt (hunk 4 above)
Comment (bpar476) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:82]: Oh I understand. Contains would not find transitive "afters"

File: spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt (hunk 6 above)
Comment (bpar476) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: I don't think it would be infinite recursion, I think it would just say that PROVISION_APPLICATION and PROVISIONING_ERROR are both after each other. Which is... kind of true 🙃

Comment (bpar476) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: Oh wait, it would cause an actual syntax error. I wonder if we could do this with method extensions after all states are declared 🤔

Comment (theghostwhoforks) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: :) yeah. The forward reference error with enums makes this annoying.

We can take a stab at the method extensions later after this PR is merged?

Comment (bpar476) [reply spi/src/main/kotlin/com/atlassian/migration/datacenter/spi/MigrationStage.kt:66]: Sure, sounds good.
